File: android/app/src/main/java/org/groestlcoin/bluewallet/MarketAPI.java
package org.groestlcoin.bluewallet;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URI;
import java.net.URL;

public class MarketAPI {

    private static final String HARD_CODED_JSON = "{\n" +
            "    \"USD\": {\n" +
            "        \"endPointKey\": \"USD\",\n" +
            "        \"locale\": \"en-US\",\n" +
            "        \"source\": \"Kraken\",\n" +
            "        \"symbol\": \"$\",\n" +
            "        \"country\": \"United States (US Dollar)\"\n" +
            "    }\n" +
            "}";

    public static String fetchPrice(String currency) {
        try {
            JSONObject json = new JSONObject(HARD_CODED_JSON);
            JSONObject currencyInfo = json.getJSONObject(currency);
            String source = currencyInfo.getString("source");
            String endPointKey = currencyInfo.getString("endPointKey");

            String urlString = buildURLString(source, endPointKey);
            URI uri = new URI(urlString);
            URL url = uri.toURL();
            HttpURLConnection urlConnection = (HttpURLConnection) url.openConnection();
            urlConnection.setRequestMethod("GET");
            urlConnection.connect();

            int responseCode = urlConnection.getResponseCode();
            if (responseCode != 200) {
                return null;
            }

            InputStreamReader reader = new InputStreamReader(urlConnection.getInputStream());
            StringBuilder jsonResponse = new StringBuilder();
            int read;
            char[] buffer = new char[1024];
            while ((read = reader.read(buffer)) != -1) {
                jsonResponse.append(buffer, 0, read);
            }

            return parseJSONBasedOnSource(jsonResponse.toString(), source, endPointKey);
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    private static String buildURLString(String source, String endPointKey) {
        switch (source) {
            case "CoinGecko":
                return "https://api.coingecko.com/api/v3/simple/price?ids=groestlcoin&vs_currencies=" + endPointKey.toLowerCase();
            default:
                return "https://api.coingecko.com/api/v3/simple/price?ids=groestlcoin&vs_currencies=" + endPointKey.toLowerCase();
        }
    }

    private static String parseJSONBasedOnSource(String jsonString, String source, String endPointKey) {
        try {
            JSONObject json = new JSONObject(jsonString);
            switch (source) {
                case "CoinGecko":
                    JSONObject bitcoinDict = json.getJSONObject("groestlcoin");
                    return bitcoinDict.getString(endPointKey.toLowerCase());
                default:
                    return null;
            }
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }
}

File: android/app/src/main/java/org/groestlcoin/bluewallet/MarketAPI.kt
package org.groestlcoin.bluewallet

import android.content.Context
import android.util.Log
import org.json.JSONObject
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL

object MarketAPI {

    private const val TAG = "MarketAPI"

    var baseUrl: String? = null

    fun fetchPrice(context: Context, currency: String): String? {
        return try {
            // Load the JSON data from the assets
            val fiatUnitsJson = context.assets.open("fiatUnits.json").bufferedReader().use { it.readText() }
            val json = JSONObject(fiatUnitsJson)
            val currencyInfo = json.getJSONObject(currency)
            val source = currencyInfo.getString("source")
            val endPointKey = currencyInfo.getString("endPointKey")

            val urlString = buildURLString(source, endPointKey)
            Log.d(TAG, "Fetching price from URL: $urlString")

            val url = URL(urlString)
            val urlConnection = url.openConnection() as HttpURLConnection
            urlConnection.requestMethod = "GET"
            urlConnection.connect()

            val responseCode = urlConnection.responseCode
            if (responseCode != 200) {
                Log.e(TAG, "Failed to fetch price. Response code: $responseCode")
                return null
            }

            val reader = InputStreamReader(urlConnection.inputStream)
            val jsonResponse = StringBuilder()
            val buffer = CharArray(1024)
            var read: Int
            while (reader.read(buffer).also { read = it } != -1) {
                jsonResponse.append(buffer, 0, read)
            }

            parseJSONBasedOnSource(jsonResponse.toString(), source, endPointKey)
        } catch (e: Exception) {
            Log.e(TAG, "Error fetching price", e)
            null
        }
    }

    private fun buildURLString(source: String, endPointKey: String): String {
        return if (baseUrl != null) {
            baseUrl + endPointKey
        } else {
            when (source) {
                "CoinGecko" -> "https://api.coingecko.com/api/v3/simple/price?ids=groestlcoin&vs_currencies=${endPointKey.lowercase()}"
                else -> "https://api.coingecko.com/api/v3/simple/price?ids=groestlcoin&vs_currencies=${endPointKey.lowercase()}"
            }
        }
    }

    private fun parseJSONBasedOnSource(jsonString: String, source: String, endPointKey: String): String? {
        return try {
            val json = JSONObject(jsonString)
            when (source) {
                "CoinGecko" -> json.getJSONObject("groestlcoin").getString(endPointKey.lowercase())
                else -> null
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error parsing price", e)
            null
        }
    }
}
